docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,229 @@
+be purely additive and not dramatically affect back-compat.
+
+### Declared Errors
+
+**Semantics:** To declare an error on an endpoint is to say “This is a very real outcome of calling this endpoint. You 
+should handle this outcome to ensure that you can react accordingly”. It should be something closely tied to the 
+functionality of that endpoint - that is to say, `getFoo` declaring `FooNotReadyYet` would be a good example, but 
+`UserNotAuthorized` would be undeclared / passed through as normal (as it is a standard expectation of interacting with 
+an authorized service).
+
+**Pros:**
+
+* Consumers now know the *expected* outcomes of calling an endpoint in a first-class way and can handle those outcomes accordingly.
+* Producers are able to express the different failure states of their endpoints in greater detail, and have conviction that those failure states are being handled by consumers.
+* Changing the failure states of an endpoint now visibly affects the API definition - changes in behavior are properly captured in code (rather than just documentation).
+
+**Cons:**
+
+* To be effective, these failure cases will need to be ‘forced’ upon consumers via the compiler. This could be as complex 
+  data types (perhaps unions of {success, error1, error2, ...} or via checked exceptions. This would entail some degree 
+  of (justified) work migrating to a with-errors world to wire each failure case into the correct behavior in client code.
+* Given the new flexibility, producers may more eagerly throw errors in their code, and ‘old’ clients will have a hard 
+  time dealing with these without upgrading.
+* Producers might be incentivized to cram *all* the possible outcomes of their endpoint as errors (see the 
+  `UserNotAuthorized` example above), and as such unduly burden consumers of their endpoint with things that should 
+  probably just be `RuntimeException`s. This in turn might have knock-on effects in the codebase depending on the exact
+  implementation used.
+* Introducing further types (whether as checked exceptions or otherwise) that determine control flow may result in them 
+  spreading throughout a code base and result in further coupling to the RPC layer. As an example, using checked exceptions
+  could encourage the consumer to throw the exception upwards rather than dealing with it at the callsite.
+  
+### Mandatory vs. Optional
+
+As discussed above, declared errors can be either made to be optional, or mandatory. Both versions come with their own
+pros and cons, and picking one side or the other inevitably involves making sacrifices.
+
+**The case for mandatory handling**
+
+_Forcing_ consumers to handle your exceptional cases is beneficial to the _producer_. You can have confidence that
+consumers will be encouraged to handle the exceptional cases, even if they are not reading your documentation. Acknowledging
+the presence of exceptional cases will be required at the minimum.
```

Comment:
> Acknowledging the presence of exceptional cases will be required at the minimum.

This is the goal based on the second paragraph

> As a consumer, you only know that an endpoint may throw an error by reading the 
documentation or the implementation itself. As a producer, you are limited in the expressivity of what you can say 
with your exception, as you have no guarantee that a consumer will handle your exception (or even read your documentation).

Reply:
The wording may be unclear - the sentence is probably better read as 'Through mandatory exceptions, users will at the very minimum have to acknowledge the existence of your error cases and consciously decide to not handle them'.

Is that what you were referring to?